src/Sulu/Bundle/WebsiteBundle/Controller/SitemapController.php

```diff
@@ -25,59 +22,137 @@
+     * @param string $alias
+     *
+     * @return Response
+     */
+    public function sitemapAction(Request $request, $alias)
+    {
+        $dumpDir = $this->getDumpDir($request);
+        if ($this->get('filesystem')->exists($dumpDir . '/sitemaps/' . $alias . '.xml')) {
+            return $this->setCacheLifetime(new BinaryFileResponse($dumpDir . '/sitemaps/' . $alias . '.xml'));
+        }
 
-        $response->headers->set(
-            HttpCache::HEADER_REVERSE_PROXY_TTL,
-            $response->getAge() + $this->container->getParameter('sulu_website.sitemap.cache.lifetime')
+        $provider = $this->get('sulu_website.sitemap.pool')->getProvider($alias);
+
+        if (1 >= ($maxPage = (int) $provider->getMaxPage())) {
+            return $this->sitemapPaginatedAction($request, $alias, 1);
+        }
+
+        return $this->setCacheLifetime(
+            $this->render(
+                'SuluWebsiteBundle:Sitemap:sitemap-paginated-index.xml.twig',
+                ['alias' => $alias, 'maxPage' => $maxPage]
+            )
         );
+    }
+
+    /**
+     * Render a single page for a single sitemap.xml provider.
+     *
+     * @param Request $request
+     * @param string $alias
+     * @param int $page
+     *
+     * @return Response
+     */
+    public function sitemapPaginatedAction(Request $request, $alias, $page)
+    {
+        $dumpDir = $this->getDumpDir($request);
+        if ($this->get('filesystem')->exists($dumpDir . '/sitemaps/' . $alias . '-' . $page . '.xml')) {
+            return $this->setCacheLifetime(
+                new BinaryFileResponse($dumpDir . '/sitemaps/' . $alias . '-' . $page . '.xml')
+            );
+        }
 
-        $response->headers->set('Content-Type', 'text/xml');
+        $portal = $request->get('_sulu')->getAttribute('portal');
+        $webspace = $request->get('_sulu')->getAttribute('webspace');
+        $localization = $request->get('_sulu')->getAttribute('localization');
 
-        $response->setContent(
-            $sitemapXMLGenerator->generate($webspaceSitemaps, $preferredDomain, $request->getScheme())
+        if (!$localization) {
+            $localization = $portal->getDefaultLocalization();
+        }
+
+        $provider = $this->get('sulu_website.sitemap.pool')->getProvider($alias);
+        $entries = $provider->generate(
+            $page,
+            $portal->getKey(),
+            $localization->getLocale()
+        );
+
+        return $this->setCacheLifetime(
+            $this->render(
+                'SuluWebsiteBundle:Sitemap:sitemap.xml.twig',
+                [
+                    'webspaceKey' => $webspace->getKey(),
+                    'locale' => $localization->getLocale(),
+                    'defaultLocale' => $portal->getXDefaultLocalization()->getLocale(),
+                    'domain' => $request->getHttpHost(),
+                    'scheme' => $request->getScheme(),
+                    'entries' => $entries,
+                ]
```

Comment:
it is maybe duplicated but  have currently no idea how to combine it ... do you have a good idea? maybe move that into an issue then we could discuss this next week?